Add new recipe for JMESPath quering of JSON

UnresolvedCold opened the pull request:
### Brief summary of what the package does

[JMESPath](https://jmespath.org/) is a quick way to query your JSON data and derive useful insights from it.
I created a small wrapper around [jamespath](https://github.com/jmespath/jp) command line utility to query a `JSON` buffer and display the queryed result in a new buffer named `*JMESPath Result*`.

### Direct link to the package repository

https://github.com/UnresolvedCold/jmespath

### Your association with the package

I am a **maintainer** and **contributor** of the package.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  Quick first pass --

- The `;;; Commentary` for this file is much wider than 80 characters (and you should wrap this at a consistent line width - downstream tools *cannot* know  where the linebreaks should go
- The summary line (top line) of the file just says "Description"
- jmespath.el#L44: Read up about `shell-quote-argument` -- the single apostrophes you're using in `"echo '%s' | jp '%s'"` will be insufficient for preventing shell control characters from wreaking havoc
- Prefer equivalent recipe: `(jmespath :repo "UnresolvedCold/jmespath" :fetcher github)` unless there's a reason not to
- jmespath.el needs *formal* license boilerplate and/or an [SPDX-License-Identifier](https://spdx.dev/ids/)

UnresolvedCold commented:
Hi @riscy 

Thanks for your review. 
I just had 1 question about licensing `jmespath.el`.
Should I incorporate the license inside the `jmespath.el` file as comments? 
Or is there any other way to license the file?

UnresolvedCold commented:
Right now, I have done the following changes. Please review and suggest me any changes that are required.  

[x] Commentary section is reduced to a few words
[x] Added a summary line at the top in place of Description
[x] Wrapped the input in `shell-quote-argument`. 
[x] Added the recipe as you said 
[x] Added a license section inside the `jmespath.el` file

FYI: @riscy

riscy commented:
As I wrote what I did below, I'm wondering: why do you need `json-mode`?  I don't see any specific dependency on it in your code.

> Should I incorporate the license inside the jmespath.el file as comments?
> Or is there any other way to license the file?

What you've got is close, but the license boilerplate usually goes after the header data.  Here's what I'd do:

```elisp
;;; jmespath.el --- Query JSON using jmespath -*- lexical-binding: t; -*-

;; Copyright (C) 2023 Shubham Kumar

;; Author: Shubham Kumar <unresolved.shubham@gmail.com>
;; Maintainer: Shubham Kumar <unresolved.shubham@gmail.com>
;; Created: December 17, 2023
;; Modified: December 17, 2023
;; Version: 0.0.1
;; Keywords: <pick one or two, not all finder-known-keywords>
;; Homepage: https://github.com/unresolvedcold/jmespath
;; Package-Requires: ((emacs "24.3") (json-mode "1.8.0"))

;; This file is not part of GNU Emacs.

;; Permission is hereby granted, free of charge, to any person obtaining a copy
;; of this software and associated documentation files (the "Software"), to deal
;; in the Software without restriction, including without limitation the rights
;; to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
;; copies of the Software, and to permit persons to whom the Software is
;; furnished to do so, subject to the following conditions:

;; The above copyright notice and this permission notice shall be included in all
;; copies or substantial portions of the Software.

;; THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
;; IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
;; FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
;; AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
;; LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
;; OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
;; SOFTWARE.

;;; Commentary:

;; A simple jmespath parser that uses jp underneath to query the json file.  All
;; you need to do is run the `jmespath-query' function and enter your jmespath
;; query.  The queried result will be displayed in the *JMESPath Result* buffer

;; This package depends on the jp utility: https://github.com/jmespath/jp
```

- Header data first, license boilerplate after
- Blank lines instead of empty `;;` lines; minor and a lot of people aren't aware of this, but this is the convention
- Headers like `;;; Installation:` actually with a triple-quote are treated like headers of new sections, and stripped from your `;;; Commentary` -- we want that text to be *part* of your commentary, so either use double semicolons for those subsections, or just remove them entirely since the content is so short.  I opted to remove them entirely
- Fixed a couple typos and help-quoted ``` `jmespath-query' ```
- Fixed the extra mention of depending on json-mode - the fact that you listed it in your `Package-Requires` is sufficient.  (But do you really depend on json-mode?)

riscy commented:
Also, you appear to have keyword-stuffed `Keywords` with every keyword in `finder-known-keywords` - choose keywords that fit.  This looks like a `Keywords: json, data` or *maybe* `Keywords: json, data, processes` sort of package.

UnresolvedCold commented:
Hi @riscy 

Thank you for your recommendations. 
I have updated the codebase as per your review. 

I was using `json-mode` to format the json generated after the query.
But it seems like this will be a unnecessary dependency on the recipe so I removed it in the latest commit.
What are your suggestion on this? Should I keep `json-mode` as a dependency for formatting the output? or will this unnecessarily force the user to install a plugin that they might not require?

riscy commented:
> I was using json-mode to format the json generated after the query.

I'm curious about how you used `json-mode` for this, because if you were using e.g. `json-pretty-print-buffer` to format the json, this is an Emacs builtin - part of json.el (*not* json-mode.el).

>  [x] Wrapped the input in shell-quote-argument.

You've run `shell-quote-argument` on only one of the two strings that's being interpolated in.  The `query` variable could also contain shell control characters.

UnresolvedCold commented:
> I'm curious about how you used json-mode for this, because if you were using e.g. json-pretty-print-buffer to format the json, this is an Emacs built-in - part of json.el (not json-mode.el).

I was changing the mode of the result buffer to `jsonc-mode` thinking it will automatically beautify my `JSON` output but that was not the case. I was mistaken. `json-mode` was not required at all as the `jp` utility already produces a beautified `JOSN` as result. 

> You've run shell-quote-argument on only one of the two strings that's being interpolated in. The query variable could also contain shell control characters.

Updated the code to add `shell-quote-argument` for query too. 
One question, is it mandatory to add `shell-quote-argument` for all the inputs which I'm using in my script?

Thank you for guiding me.
Updated commit - https://github.com/UnresolvedCold/jmespath/commit/18069adbf47b934cb1fb475085f1ee0069684c79

riscy commented:
> is it mandatory to add shell-quote-argument for all the inputs which I'm using in my script?

Yes, when I see it (and I don't always catch it in time) I'll let maintainers know they absolutely want to quote user-controllable inputs to the shell.  What if the user accidentally ran `jmespath-query` on a buffer that wasn't valid json, but was a scratch buffer or a shell script of some kind?  What if the user's query involved shell control characters but was malformed?  Arbitrarily bad things could happen.

Zooming back out to talk about the package itself, this is relatively small (a single function, focused on a single use-case).  I think there's value here, but maybe it's not ready yet for MELPA.  What *would* make it ready for MELPA would be a more comprehensive treatment of what it looks like jp can do.

- For example, you've written a `jmespath-query` but I think you should call this `jmespath-query-buffer`.  (I'd personally be tempted to give it an optional BUFFER argument (e.g. `(defun jmespath-query-buffer (query &optional buffer) ...)` that would make it a little easier for people to target a specific buffer)
- You could write a `jmespath-query-file` to support jp's ability to read from files: https://github.com/jmespath/jp#input-from-a-file
- Instead of popping to a buffer with the result, which really makes your entire package useable only in an interactive way (and not useable by other programs), you should support the functionality of just returning the string the user wants.  If I wanted to use this as a dependency of one of my packages, I'd have to write code to parse the contents of the `*JMESPath Result*` buffer.  I think just returning the string is far more useful in general, but to continue supporting the interactive use case you could have two functions, e.g. `jmespath-query-buffer-and-show` and `jmespath-query-buffer` (for example).  The former could be `(interactive)`, while the latter would comprise your programmatic API.

UnresolvedCold commented:
Thanks for the feedback @riscy 
I will incorporate these changes soon.

UnresolvedCold commented:
Hi @riscy 

I have updated the recipe with the following changes.

1. Updated function `jmespath-query-buffer`. This can now take an optional buffer as an argument and return a string output as result.
2. New functin `jmespath-query-file`. This function takes a query and filename as input and returns the result as string.

UnresolvedCold commented:
Hi @riscy 

I have updated the recipe with the following changes.

1. Updated function `jmespath-query-buffer`. This can now take an optional buffer as an argument and return a `string` output as result.
2. New function `jmespath-query-file`. This function takes a query and filename as input and returns the result as `string`.
3. New function 'jmespath-query-and-show' *(interactive)* can take the current buffer or a file as input and will display the result in `*JMESPath Result*` buffer. `C-u M-x jmespath-query-and-show` will ask for a filename as input while `M-x jmespath-query-and-show` will act on the current buffer.

riscy commented:
Looks good - thank you for accommodating!

I won't block on this, but the commentary for `jmespath-query-and-show` doesn't explain to users that `current-prefix-arg` is used to determine whether the current buffer or a supplied filename will be asked for.  (I'd also rename that variable from INPUT to TARGET, perhaps, but that's just due to my own subjective sense of naming.)

riscy commented:
And, welcome to MELPA. :)
